JoinColony/colonyCDapp
# Fix Reputation Cache Update Lambda

This PR fixes the way the `updateContributorsWithReputation` Lambda updates the cached reputation for users by shedding it's the network _(and implicitly `colonyJS`)_, making only the db and the Reputation miner the only parts required for it to work.

The initial issue that triggered this was when a block ingestor fell behind the network, and when it caught up, and processed the `ReputationMiningCycleComplete` event, it used the wrong timestamp and delayed the reputation for the whole app for at least one mining cycle.

This was facilitated by the introduction of a `/rootHashes` route in the miner on the network side, which will return the current root hash as well as the upcoming one https://github.com/JoinColony/colonyNetwork/commit/3f8df23173f840226fb8da42547e12664f10c290

Using the above endpoint we can now generate everything we need inside the lambda using simple fetches directly to the miner, without the need to instantiate `colony-js` clients or rpc endpoints 

On the ingestor side, this fixes the way the timestamp is being saved for the block ingestor reputation miner update event, by fetching the actual block one, instead of the `Date.now()` timestamp. This corrects a issue that arises when the block ingestor falls behing the chain then catches up JoinColony/block-ingestor#296

Since we had to update to the latest network release, we also had to update `colony-js` to it's latest `next` release both in the CDapp, Lambdas, as well as the Block Ingestor https://github.com/JoinColony/colonyNetwork/releases/tag/imwss3 

Due to the above update, all the versions have now been updated, both colonies as well as exptensions

![Screenshot from 2024-12-03 16-29-53](https://github.com/user-attachments/assets/6f30dead-a9fc-4e1d-b144-88ddcc433a9e)
![Screenshot from 2024-12-03 16-29-32](https://github.com/user-attachments/assets/1d923a3a-3541-466c-bca5-e1fec1decf91)

## Future

This is just a temporary improvement, as the way we're updating the reputation currently is not scalable, meaning that on colonies with a lot of contributing users (users with reputation), this will simply fail to update everyone's reputation as there's a hard cap on execution time.

This will require re-engineering and re-architecturing to solve

## Testing

1. Run the dev env
2. Run the `create-data` script
3. Run the frontent
4. Go to `planex`
5. Load up the Members page and/or the Teams page and make a mental note of the reputation values
![Screenshot from 2024-12-03 16-15-10](https://github.com/user-attachments/assets/f2bcd106-8bba-4acf-beb6-f499a8b0a8f9)
![Screenshot from 2024-12-03 16-15-17](https://github.com/user-attachments/assets/a2f1d9b3-bf2d-4757-a913-d6838f80072a)
6. Apply this patch _(which basically reverts the lambda to it's old state, and forces a cache update on each lambda call)_ [reputation-cache-update.patch](https://github.com/user-attachments/files/17996284/reputation-cache-update.txt)
7. Make sure to run `npm i` again after applying the patch so that the new _(old)_ dependencies get re-installed in the `updateContributorsWithReputation` lambda
8. Go back to `planex`, and refresh the page _(if you already didn't)_
9. Go back to the Members and/or Teams pages and confirm that the values are the same 

This proves that the no reputation logic changes happen, and the change here is just to do with the place we get those values

_NOTE: If you are lucky enough (or unlucky), you might be testing this just in the middle of the reputation period happening, so when comparing values, you will get different (lower) values. If that happens, just revert and compare again using the current PRs changes to the lambda_

JoinColony/block-ingestor#296
